provider-forums/src/java/org/sakaiproject/archiver/provider/ForumsArchiver.java

```diff
@@ -0,0 +1,394 @@
+import org.sakaiproject.archiver.api.ArchiverRegistry;
+import org.sakaiproject.archiver.api.ArchiverService;
+import org.sakaiproject.archiver.spi.Archiveable;
+import org.sakaiproject.archiver.util.Jsonifier;
+import org.sakaiproject.content.api.ContentHostingService;
+import org.sakaiproject.exception.IdUnusedException;
+import org.sakaiproject.exception.PermissionException;
+import org.sakaiproject.exception.ServerOverloadException;
+import org.sakaiproject.exception.TypeException;
+
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Implementation of {@link Archiveable} for various content tools
+ */
+@Slf4j
+public class ForumsArchiver implements Archiveable {
+
+	private static final String TOOL_ID = "sakai.forums";
+
+	@Setter
+	protected DiscussionForumManager forumManager;
+
+	public void init() {
+		ArchiverRegistry.getInstance().register(TOOL_ID, this);
+	}
+
+	public void destroy() {
+		ArchiverRegistry.getInstance().unregister(TOOL_ID);
+	}
+
+	@Setter
+	private ContentHostingService contentHostingService;
+
+	@Setter
+	private ArchiverService archiverService;
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public void archive(final String archiveId, final String siteId, final String toolId, final boolean includeStudentContent) {
+
+		final List<DiscussionForum> forums = this.forumManager.getDiscussionForumsWithTopics(siteId);
+
+		for (final DiscussionForum forum : forums) {
+			// Archive the forum, including topics (and messages if includeStudentContent is true)
+			archiveForum(forum, archiveId, siteId, toolId, includeStudentContent);
+
+			// Archive the attachments for this forum
+			archiveAttachments(forum.getAttachments(), "forum: " + forum.getTitle(), forum.getTitle() + "/forum-attachments/", archiveId,
+					siteId, toolId);
+		}
+	}
+
+	/**
+	 * Archive the topics within a forum
+	 *
+	 * @param forum
+	 * @param archiveId
+	 * @param siteId
+	 * @param toolId
+	 */
+	@SuppressWarnings("unchecked")
+	private void archiveForum(final DiscussionForum forum, final String archiveId, final String siteId, final String toolId,
+			final boolean includeStudentContent) {
+
+		// Set up the simple forum object
+		final SimpleForum simpleForum = new SimpleForum(forum);
+
+		// Initialise the array to hold each topic for this forum
+		final List<SimpleTopic> simpleTopics = new ArrayList<SimpleTopic>();
+
+		final List<DiscussionTopic> topics = forum.getTopics();
+		for (final DiscussionTopic topic : topics) {
+			final SimpleTopic simpleTopic = new SimpleTopic(topic);
+
+			// Add this topic to the forum's topic array
+			simpleTopics.add(simpleTopic);
+
+			// Set up the folder structure for saving attachments and messages
+			final String folderStructure = forum.getTitle() + "/topics/" + topic.getTitle();
+
+			// Archive the attachments for this topic
+			final String errorTopicAttachment = "topic: " + topic.getTitle() + " in forum: " + forum.getTitle();
```

Comment:
Because I made it so that forum, topic and message attachments all use the archiveAttachments method, I needed a way to pass in different error messages for each of them. So that is the error message for topic attachments. Suggestions welcome.